ast/node.go

```diff
@@ -0,0 +1,523 @@
+// {echocmd}: echo, echon, echomsg, echoerr
+type EchoCmd struct {
+	Start   Pos    // position of starting the echo-command
+	CmdName string // echo-command name
+	ExArg   ExArg  // Ex command arg
+	Exprs   []Expr
+}
+
+func (f *EchoCmd) Pos() Pos { return f.Start }
+func (f *EchoCmd) Cmd() Cmd { return *f.ExArg.Cmd }
+
+// vimlparser: ECHOHL .ea .str
+// :echohl {name}
+type Echohl struct {
+	Echohl Pos   // position of starting the :echohl
+	ExArg  ExArg // Ex command arg
+	Name   string
+}
+
+func (f *Echohl) Pos() Pos { return f.Echohl }
+func (f *Echohl) Cmd() Cmd { return *f.ExArg.Cmd }
+
+// vimlparser: EXECUTE .ea .list
+// :execute {Expr}..
+type Execute struct {
+	Execute Pos   // position of starting the :execute
+	ExArg   ExArg // Ex command arg
+	Exprs   []Expr
+}
+
+func (f *Execute) Pos() Pos { return f.Execute }
+func (f *Execute) Cmd() Cmd { return *f.ExArg.Cmd }
+
+// vimlparser: TERNARY .cond .left .right
+// Condition ? Left : Right
+type TernaryExpr struct {
+	Ternary   Pos // position of starting the :execute
+	Condition Expr
+	Left      Expr
+	Right     Expr
+}
+
+func (f *TernaryExpr) Pos() Pos { return f.Ternary }
+
+type BinaryExpr struct {
+	Left  Expr        // left operand
+	OpPos Pos         // position of Op
+	Op    token.Token // operator
+	Right Expr        // right operand
+}
+
+func (f *BinaryExpr) Pos() Pos { return f.OpPos }
+
+type UnaryExpr struct {
+	OpPos Pos         // position of Op
+	Op    token.Token // operator
+	X     Expr        // operand
+}
+
+func (f *UnaryExpr) Pos() Pos { return f.OpPos }
+
+// Left[Right]
+type SubscriptExpr struct {
+	Lbrack Pos // position of "["
+	Left   Expr
+	Right  Expr
+}
+
+func (f *SubscriptExpr) Pos() Pos { return f.Lbrack }
+
+// X[Low:High]
+type SliceExpr struct {
+	X      Expr // expression
+	Lbrack Pos  // position of "["
+	Low    Expr // begin of slice range; or nil
+	High   Expr // end of slice range; or nil
+}
+
+func (f *SliceExpr) Pos() Pos { return f.Lbrack }
+
+// vimlparser: CALL .left .rlist
+type CallExpr struct {
+	Fun    Expr   // function expression
+	Lparen Pos    // position of "("
+	Args   []Expr // function arguments; or nil
+}
+
+func (c *CallExpr) Pos() Pos { return c.Lparen }
+
+// Left.Right
+// vimlparser: Dot .left .right
+type DotExpr struct {
+	Left  Expr
+	Dot   Pos // position of "."
+	Right Ident
+}
+
+func (c *DotExpr) Pos() Pos { return c.Dot }
+
+type BasicLit struct {
+	ValuePos Pos         // literal position
+	Kind     token.Token // token.INT, token.STRING, token.OPTION, token.ENV, token.REG
+	Value    string
+}
+
+func (c *BasicLit) Pos() Pos { return c.ValuePos }
+
+type List struct {
+	Lsquare Pos // position of "["
+	Values  []Expr
+}
+
+func (c *List) Pos() Pos { return c.Lsquare }
+
+type Dict struct {
+	Lcurlybrace Pos // position of "{"
+	Entries     []KeyValue
+}
+
+func (c *Dict) Pos() Pos { return c.Lcurlybrace }
+
+type KeyValue struct {
+	Key   BasicLit // string
+	Value Expr
+	// TODO: want Pos data...
```

Comment:
it might be enough to get pos of key and value position. `:` should be right after the key.